src/measurements.rs

```diff
@@ -42,47 +42,62 @@ where
 /// Returns a new mask, containing the biggest zone of `mask`.
 ///
 /// * `mask` - Binary image to be labeled and studied.
-pub fn largest_connected_components<S>(mask: &ArrayBase<S, Ix3>) -> Option<Mask>
+/// * `structure` - Structuring element used for the labeling. Must be 3x3x3 (e.g. the result
+///   of [`Kernel3d::generate`](crate::Kernel3d::generate)) and centrosymmetric. The center must be `true`.
+pub fn largest_connected_components<S>(
+    mask: &ArrayBase<S, Ix3>,
+    structure: &ArrayBase<S, Ix3>,
+) -> Option<Mask>
 where
     S: Data<Elem = bool>,
 {
-    let (labels, nb_features) = label(mask);
+    let (labels, nb_features) = label(mask, structure);
     let (right_label, _) = most_frequent_label(&labels, nb_features)?;
     Some(labels.mapv(|l| l == right_label))
 }
 
 /// Labels features of 3D binary images.
 ///
-/// Currently hardcoded with the Star kernel (`Kernel3d::Star`).
-///
 /// Returns the labels and the number of features.
 ///
 /// * `mask` - Binary image to be labeled. `false` values are considered the background.
-pub fn label<S>(data: &ArrayBase<S, Ix3>) -> (Array3<u16>, usize)
+/// * `structure` - Structuring element used for the labeling. Must be 3x3x3 (e.g. the result
+///   of [`Kernel3d::generate`](crate::Kernel3d::generate)) and centrosymmetric. The center must be `true`.
+pub fn label<S>(data: &ArrayBase<S, Ix3>, structure: &ArrayBase<S, Ix3>) -> (Array3<u16>, usize)
 where
     S: Data<Elem = bool>,
 {
+    assert!(structure.shape() == &[3, 3, 3], "`structure` must be size 3 in all dimensions");
+    assert!(structure == structure.slice(s![..;-1, ..;-1, ..;-1]), "`structure is not symmetric");
+
     let len = data.dim().2;
     let mut line_buffer = vec![BACKGROUND; len + 2];
     let mut neighbors = vec![BACKGROUND; len + 2];
 
     let mut next_region = FOREGROUND + 1;
     let mut equivalences: Vec<_> = (0..next_region).collect();
 
-    // We only handle 3D data with a 6-connectivity kernel for now, but this algo can handle
-    // N-dimensional data and any symmetric kernels.
+    // We only handle 3D data for now, but this algo can handle N-dimensional data.
     // https://github.com/scipy/scipy/blob/v0.16.1/scipy/ndimage/src/_ni_label.pyx
     // N-D: Use a loop in `is_valid` and change the `labels` indexing (might be hard in Rust)
-    // Kernels: Add new `kernel_data` and choose the right one
-    let kernel_data = [
-        //((false, false, false), [0, 0]),
-        ([false, true, false], [0, 1]),
-        //((false, false, false), [0, 0]),
-        ([false, true, false], [1, 0]),
-    ];
-    let nb_neighbors = kernel_data.len();
 
-    let use_previous = true;
+    let nb_neighbors = structure.len() / 6;
+    let kernel_data: Vec<([bool; 3], [isize; 2])> = structure
+        .lanes(Axis(2))
+        .into_iter()
+        .zip(0isize..nb_neighbors as isize)
```

Comment:
If this is the laneindex, can't you use `0isize..` instead? `zip` will stop when it needs to.

Reply:
I'm being a little sneaky here and using `zip` as a combined `enumerate` and `take(nb_neighbors)`. The `enumerate` part was just so that I could get the index as a `isize` instead of a `usize`, to avoid lots of `as isize` below. And the `take` part is used to remove half of the lanes, as they are symmetric to the half we use.

I can chop it up into an explicit `.zip(0isize...).take(nb_neighbors)` if you want.

Reply:
Sorry, I was unclear. You're zipping with `structure.lanes(Axis(2))` and `0isize..nb_neighbors as isize`.
- Are they always the same size? If yes, then you should be able to use `0isize..` because `zip` stops at the shorter iterator.
- If not, then which one is longer?

Reply:
The length of is `structure.lanes(Axis(2))` is `structure.len() / 3`, while `nb_neighbors` is `structure.len() / 6`, so the range is always half the length of the lanes. Since it's not obvious, I'll switch it to use `take` instead.